website/project/views/contributor.py

```diff
@@ -752,8 +752,10 @@ def claim_user_form(auth, **kwargs):
             # Authenticate user and redirect to project page
             status.push_status_message(language.CLAIMED_CONTRIBUTOR, kind='success', trust=True)
             # Redirect to CAS and authenticate the user with a verification key.
+            provider = PreprintProvider.load(pid)
+            redirect_url = web_url_for('resolve_guid', guid=pid, _absolute=True) if provider is None else provider.landing_url
```

Comment:
I had the same concern, but came to the same realization. On the bright side, emberizing the front-end will eventually obviate these views.

Reply:
@fabmiz  CC, @mfraezz and @sloria 

For CAS redirect in this case, it uses the `username` and `verification_key` for login. This requires the base redirect URL to be an OSF URL that knows auth backend in order to handle the authentication first. Otherwise, the auth is left in a state where CAS is logged in but OSF is not. That's why when you click sign in again, users are automatically logged in.

For URLs that do not know the backend auth logic (e.g. preprints, registries, etc.), we use `https://osf.io/login/?next=destination`.

For example, preprints login URL is

```
https://accounts.osf.io/login?service=
    https%3A%2F%2Fosf.io%2Flogin%2F%3Fnext%3D
        https%253A%252F%252Fosf.io%252Fpreprints%252F
```

For the matter here, this will work.

```python
if provider:
    redirect_url = web_url_for('auth_login', next=provider.landing_url, _absolute=True)
else
   ...
```

There is a whitelist check for the URL given to `next`.